File: exam/string/valid_palindrome/valid_palindrome.dart
void main() {
  String name = "A man, a plan, a camel: Panama";
  String name2 = "A man, a plan, a canal: Panama";

  String result = validString(name);
  String result2 = validString(name2);
  print(result);
  print(result.isPalindrome);
  print(result2.isPalindrome);
}

String validString(String name) {
  List<String> temp = [];

  List<String> data = name.trim().split('').toList();
  data.forEach((element) {
    if (validChar(element)) {
      temp.add(element.toLowerCase());
    }
  });

  String data2 = temp.map((e) => e).join('');

  return data2;
}

bool validChar(String name) {
  if ((name.codeUnitAt(0) >= "A".codeUnitAt(0) &&
          name.codeUnitAt(0) <= "Z".codeUnitAt(0)) ||
      (name.codeUnitAt(0) >= "a".codeUnitAt(0) &&
          name.codeUnitAt(0) <= "z".codeUnitAt(0)) ||
      (name.codeUnitAt(0) >= "0".codeUnitAt(0) &&
          name.codeUnitAt(0) <= "9".codeUnitAt(0))) {
    return true;
  }
  return false;
}

bool isPalindromeString(String name) {
  return name.split('').toList().reversed == name;
}

extension StringUtilities on String {
  // Method to reverse a string
  String get reversed {
    return this.split('').reversed.join('');
  }

  // Method to check if the string is a palindrome
  bool get isPalindrome {
    return this.toLowerCase() == this.reversed.toLowerCase();
  }
}

File: exam/string/valid_palindrome/valid_palindrome.cpp
#include <iostream>
using namespace std;

char toLowerCase(char ch)
{
    if ((ch >= 'a' && ch <= 'z') || (ch == '0' && ch == '9'))
    {
        return ch;
    }
    char temp = ch - 'A' + 'a';
    return temp;
}

bool isValidChar(char ch)
{

    if ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'z') || (ch == '0' && ch == '9'))
    {
        return 1;
    }
    return 0;
}

bool isPalindrome(string name)
{

    int s = 0;
    int e = name.length() - 1;

    while (s <= e)
    {
        if (name[s] != name[e])
        {
            return 0;
        }
        else
        {
            s++;
            e--;
        }
    }

    return true;
}

string validString(string name)
{
    string temp = "";
    for (int i = 0; i < name.length(); i++)
    {
        if (isValidChar(name[i]))
        {
            temp.push_back(toLowerCase(name[i]));
        }
    }

    cout << temp << endl;

    return temp;
}

int main()
{

    // Leet Code : 125
    string name = "A man, a plan, a camel: Panama";
    string name2 = "A man, a plan, a canal: Panama";

    cout << isPalindrome((validString(name))) << endl;
    cout << isPalindrome((validString(name2))) << endl;

    return 0;
}